### ISSUE TITLE
mutate_gaussian handling of no assigned std

### ISSUE THREAD

**Issue by grudloff** [AureumChaos/LEAP#111]
If `mutate_gaussian` is not assigned an `std` it leads to a `ValueError` because it doesn't return an iterator which is somewhat confusing for a first-timer. Maybe it could be a good idea to check it or to have a default value?

**Comment by piprrr** [comment]
Thanks for bringing this to our attention!

@SigmaX , if I run `examples/simple_es.py` it now fails, which is related to this issue:

```
  File "/Users/may/Projects/LEAP/leap_ec/real_rep/ops.py", line 45, in mutate_gaussian
    raise ValueError("No value given for expected_num_mutations.  Must be either a float or the string 'isotropic'.")
ValueError: No value given for expected_num_mutations.  Must be either a float or the string 'isotropic'.
```

Just pinging you because I know you were making another pass over the `expected_num_mutations` vs. `mutation probability` conundrum that we were hand wringing over.  Maybe we should have this parameter default to '1' or `isotropic`?  Also, what do you think of putting a `*` to force the use of keyword arguments?  I was thinking of doing this since we use `@curry`, which will cause using positional arguments to fail; keyword arguments must be used, or the `@curry` will fail.  (I had this happen with someone else, so this isn't the first time..)

**Comment by SigmaX** [comment]
Good catch; thanks @grudloff!  This sounds like a case where we need to add a descriptive error.

@piprrr I chose to remove default parameters from `expected_num_mutations` and `probability` because in my own experience I found myself prone to making incorrect assumptions about what the default value is (i.e. I've run big batches of experiments thinking it was 1/L when it was actually isotropic, and vice-versa).

The same reasoning applies here, so I lean toward adding a descriptive `ValueError` rather than a default value.

Good point about `@toolz.curry`.  Its propensity to generate  operators that produce cryptic errors has worried me.   We might try out the `*` trick and see how we like it.

**Comment by SigmaX** [comment]
This also affects the `std` parameter of `leap_ec.int_rep.mutate_binomial()`, and in fact any operator argument that does not have a default value (our various `bounds` arguments, for instance).

The **general problem** is that we use `@toolz.curry` to conveniently write operators as single functions, where users can "set all parameter except the `population`," producing a simple function `op(population)` that applies to populations.  But this also means that **partial currying** is possible, where the user sets some parameters, but not others, leaving a function like `op(population, std)`, which pops out a function `op(std)` when the pipeline is run (instead of a modified `population` iterator).

Ergo the exception in the downstream operator:

```
ValueError: Operator <function evaluate at 0x7fae35830c10> received a <class 'toolz.functoolz.curry'> as input, but expected an iterator.
```

 - One fix is to set a default value for all operator parameters, i.e. `std: float = None`, and insert exception handling for every parameter to throw a `ValueError` if the parameter is omitted:

  ```
  ValueError: The `std` parameter is `None`, but must be specified.
  ```

But we would have to do that for every operator argument with no default value, always.  That's a lot of work for something that is handled automatically by the compiler in the case of closures or classes.

 - I fiddled with `*` for a bit, but I don't think it offers us any way to prohibit partial currying.

This leaves me with two ideas:

 - Abandon `@toolz.curry` entirely.  It's convenient, but closures are conveneint too, and closures don't have the partial currying problem.

 - Write some kind of decorator that can detect when a user forgets a required argument and raises a `ValueError`, much like we use `@iteriter_op` and friends to ensure that operators accept & output the correct types.

@piprrr, do you have any other ideas?

**Comment by SigmaX** [comment]
Made the error message more descriptive.  This doesn't pinpoint exactly _which_ operator/argument failed to be provided, but at least gives the user a strong indication of what the problem is:

```
ValueError: While executing operator <function evaluate at 0x7f5383bf5b80>, an incomplete curry object was received (<class 'toolz.functoolz.curry'>).
This usually means that you forgot to specify a required argument for an upstream operator, so a partly-curried function got passed down the pipeline instead of a population iterator.
```

In `develop`, to be released with the upcoming `v0.6.0`.
